### PR TITLE
feat: Use encrypted cookie to store OAuth2 state nonce (instead of redis)

### LINKED ISSUE #5873: Use encrypted cookie to store OAuth2 state nonce (instead of redis)
**Describe the bug**

Currently (as of v2.0) during the OAuth2 exchange, we store the state nonce in redis, which we retrieve to complete the exchange.

When redis is not functioning (e.g. because of an update), a user that attempts login will get the error: `unknown app state <oidc state nonce>`

Argo Workflows has developed a technique where the state nonce is stored as an encrypted cookie, rather than in redis, which allows them to avoid the use of redis entirely during SSO login.

https://github.com/argoproj/argo-workflows/pull/2745

The following post describes the technique:

https://security.stackexchange.com/questions/140883/is-it-safe-to-store-the-state-parameter-value-in-cookie

**To Reproduce**

Shut down redis, attempt an SSO login

**Expected behavior**

SSO login flow should still work without redis

**Version**

v2.0

### PR DESCRIPTION
**Pull request by alexmt**
Signed-off-by: Alexander Matyushentsev <AMatyushentsev@gmail.com>

Closes https://github.com/argoproj/argo-cd/issues/5873

PR changes SSO auth flow to store temporary nonce tokens in an encrypted cookie instead of redis

### CODE REVIEW

File: util/oidc/oidc.go
Comment (jannfis) [inline util/oidc/oidc.go:173]: We should make sure that `len(data) > nonceSize` to prevent out-of-bounds access to `data` slice, even if that might be only remotely possible. We should not just assume that `data` will be of the same size to the `nonceSize` or larger.

File: util/oidc/oidc.go
Comment (jannfis) [inline util/oidc/oidc.go:144]: I'm not sure if we should use `md5` here to derive the key from the passphrase. From the example of the Go docs at https://pkg.go.dev/crypto/cipher#NewGCM, they stress that a cryptographic secure mechanism should be used to derive the key from the passphrase, such as `bcrypt` or `scrypt`. 

If we decide to chose a one-way, non-salted hashing algorithm, I think it should be at least SHA-256 instead of MD5. This would also make the key 32 byte long and use AES-256 instead of AES-128. If we switch to SHA-256, also the follow-up `[]byte(hex.EncodeToString(...))` wouldn't be necessary, and we could just supply `hasher.Sum(nil)` to `aes.NewCipher()` as key.

**Review by jannfis — CHANGES_REQUESTED** [review]
From the first glance, I have a few questions and challenges to discuss.

Since we should be careful with this change, I'm putting a request for change for now.

File: util/oidc/oidc.go
Comment (leoluz) [inline util/oidc/oidc.go:142]: What do you think about having the encrypt/decrypt as global utility functions that could be reused in different parts of the application?

Comment (alexmt) [reply util/oidc/oidc.go:142]: Agree. Moving it to `util/crypto`

**Comment by alexmt** [comment]
Thank you for the review, @leoluz  and @jannfis !

@jannfis do you think that change is too dangerous and should be postponed to v2.4?

**Comment by jannfis** [comment]
> @jannfis do you think that change is too dangerous and should be postponed to v2.4?

It's not dangerous per se, but we should carefully review the code, flow of data and possible pitfalls before merging. Change itself is fine to go into 2.3 IMHO.

File: util/oidc/oidc.go
Comment (alexmt) [reply util/oidc/oidc.go]: done

File: util/oidc/oidc.go
Comment (alexmt) [reply util/oidc/oidc.go]: I really like that we can get rid of `[]byte(hex.EncodeToString(...))` with SHA-256 so I've chosen this option

**Comment by alexmt** [comment]
Thanks for review, @jannfis  and @leoluz . Your comments have been addressed. PTAL

File: util/crypto/crypto.go
Comment (jannfis) [inline util/crypto/crypto.go:11]: Maybe rename this function to `passphraseToKey` or `keyFromPassphrase`?

File: util/crypto/crypto.go
Comment (jannfis) [inline util/crypto/crypto.go:47]: Can we please add a check for `len(data) > nonceSize` before we access the slice in this way?

File: util/oidc/oidc.go
Comment (jannfis) [inline util/oidc/oidc.go:154]: Why do we set the 3 minutes expiry time?

File: util/oidc/oidc.go
Comment (jannfis) [inline util/oidc/oidc.go:190]: Is this supposed to unset the cookie? Or why do provide a value of `""` here?

**Review by jannfis — COMMENTED** [review]
Few more comments.

File: util/oidc/oidc.go
Comment (crenshaw-dev) [inline util/oidc/oidc.go:180]:
```suggestion
	if len(parts) == 2 {
		returnURL = parts[1]
	} else {
		return "", fmt.Errorf("invalid format in '%s' cookie", common.AuthCookieName)
	}
```

It should never happen. But if I were really eccentric and encrypted my own cookie, I could get Argo CD to redirect me to an empty location, which has [uncertain behavior](https://snook.ca/archives/servers/empty_location).

Comment (jannfis) [reply util/oidc/oidc.go:180]: Good catch. I think an implicit check should also be for `parts[1] != ""`, because input string `foo:` would also yield in a len of `parts` of 2, but with an empty second slice item, e.g. `if len(parts) == 2 && parts[1] != "" ...`

File: util/oidc/oidc.go
Comment (alexmt) [reply util/oidc/oidc.go:191]: Correct. This is a way to remove cookie. Added comment

File: util/oidc/oidc.go
Comment (alexmt) [reply util/oidc/oidc.go]: Thank you! Added a check and unit test

File: util/oidc/oidc.go
Comment (alexmt) [reply util/oidc/oidc.go]: That cookie is required only during login process so we don't want to store it for a long time. Move timeout to a constant with the comment.

File: util/crypto/crypto.go
Comment (alexmt) [reply util/crypto/crypto.go]: Sorry, did not notice it in first comment; fixed.

File: util/crypto/crypto.go
Comment (alexmt) [reply util/crypto/crypto.go]: Sounds good. Done

**Comment by alexmt** [comment]
Thanks for the careful review! All comments have been addressed. PTAL

File: util/crypto/crypto.go
Comment (crenshaw-dev) [inline util/crypto/crypto.go:16]:
```suggestion 
	return sha256.Sum256([]byte(passphrase))
```

Could you get away with this? Looks like the hasher might be more designed for when you have chunks of data to add to the hash.

File: util/crypto/crypto.go
Comment (crenshaw-dev) [inline util/crypto/crypto.go:33]: What's the role of the `nonce` as first param here?

File: util/crypto/crypto.go
Comment (crenshaw-dev) [inline util/crypto/crypto.go:13]: Should we use bcrypt instead of sha256? It's [recommended](https://pkg.go.dev/crypto/cipher#example-NewGCM-Encrypt) in the crypto docs for keys.

Since the key is 32 random bytes, I don't think bcrypt actually buys us much. But if someone found a flaw in the random number generator, bcrypt would give us just a tiny bit more protection against someone brute-forcing the hash to recover the key.

**Review by crenshaw-dev — CHANGES_REQUESTED** [review]

File: util/crypto/crypto.go
Comment (crenshaw-dev) [reply util/crypto/crypto.go:33]: Ah nevermind, I see that the ciphertext gets appended to the nonce, so you have access to the nonce in Decrypt.

File: util/oidc/oidc.go
Comment (crenshaw-dev) [inline util/oidc/oidc.go:181]: Should there be a check for len(parts) == 0?

File: util/oidc/oidc_test.go
Comment (crenshaw-dev) [inline util/oidc/oidc_test.go:190]: Should we use a password generated by `MakeSignature`, just so it matches the use case more closely?

otoh I can see the downside of introducing random input for a non-fuzzing test.

File: util/oidc/oidc.go
Comment (crenshaw-dev) [inline util/oidc/oidc.go:171]: Absurdly tiny nitpick: could we just pass around the signature as a byte slice rather than converting to/from string? Benefits are: 1) less likely someone will accidentally log it and 2) fewer copies of the secret in memory (very very very tiny benefit).

File: util/crypto/crypto.go
Comment (crenshaw-dev) [reply util/crypto/crypto.go:13]: Actually... why hash at all? The server token is already 32 random bytes. Couldn't it serve directly as the key? Since this hashed key disappears at GC, seems like it's functionally the same as just using the server token directly.

Comment (crenshaw-dev) [reply util/crypto/crypto.go:13]: (I'm assuming we don't let users set that secret to some arbitrary string.)

File: util/oidc/oidc.go
Comment (alexmt) [reply util/oidc/oidc.go]: done

File: util/oidc/oidc_test.go
Comment (alexmt) [reply util/oidc/oidc_test.go]: done

File: util/crypto/crypto.go
Comment (alexmt) [reply util/crypto/crypto.go]: 🤦 I should have look at how server signature is generated. Removed hashing completelly

File: util/crypto/crypto.go
Comment (alexmt) [reply util/crypto/crypto.go]: Removed hashing since ServeSignature is already 32 random bytes

File: util/oidc/oidc.go
Comment (jannfis) [reply util/oidc/oidc.go:189]: `strings.Split()` and `strings.SplitN()` always return a slice containing at least 1 element.

File: util/crypto/crypto.go
Comment (jannfis) [reply util/crypto/crypto.go]: I was thinking: What if we decided at some point in time to make the server's signature different from what it is now (e.g. smaller or larger, different format, etc)? Also, I think now we have made the `Encrypt()` and `Decrypt()` functions public in our package, but they are of limited use now, i.e. only usable with our server's signature. SHA256 in contrast has a constant output of 32 bytes, for any given passphrase, even if it comes at some additional computational cost.

File: util/crypto/crypto.go
Comment (crenshaw-dev) [inline util/crypto/crypto.go:23]: Why panic instead of just returning the error?

Comment (alexmt) [reply util/crypto/crypto.go]: fixed

**Review by crenshaw-dev — APPROVED** [review]
LGTM

**Review by jannfis — APPROVED** [review]
LGTM

File: util/crypto/crypto.go
Comment (alexmt) [reply util/crypto/crypto.go]: Sorry resolved this thread before I noticed your comment @jannfis . I think you are right. I don't want to accidentally break SSO tomorrow. I've fallback to original @crenshaw-dev 's suggestion: use scrypt to generate keys from the keyphrase (bcrypt does not support generating 32 bytes long keys). According to docs, it is actually very memory expensive so I changed the code to generate key only once when settings are loaded/updated. Now crypto package has Encrypt/Decrypt functions that can be used with any key and function that generates encryption key using server signature.

**Comment by alexmt** [comment]
I've noticed that one conversation was resolved incorrectly, unresolved it, and made some improvements :) . Please see https://github.com/argoproj/argo-cd/pull/8241#discussion_r792166406 for more information.

Sorry for moving back and forth in this PR. Cryptography is definitely not my strength and it was a learning exercise for me.

File: util/crypto/crypto.go
Comment (crenshaw-dev) [inline util/crypto/crypto.go:20]: Is the salt actually helpful for this use case, or is it functionally equivalent to no-salt?

**Review by crenshaw-dev — COMMENTED** [review]
No comments that should block merging. I have a _slight_ preference for explicitly using a "nothing" salt if the hashed passphrase isn't actually a meaningful salt.

File: util/crypto/crypto.go
Comment (alexmt) [reply util/crypto/crypto.go:20]: As I understand it is equivalent of no salt. I think it would be overkill to generate and store salt in argocd-secret Secret just to derive key from server signature

Comment (crenshaw-dev) [reply util/crypto/crypto.go:20]: Cool, agreed. For readability, what about explicit no salt (empty slice?) or a comment clarifying why we’re using the salt value we’re using.

**Review by jannfis — APPROVED** [review]
LGTM, thanks for the extra mile on this @alexmt

File: util/crypto/crypto.go
Comment (alexmt) [reply util/crypto/crypto.go:21]: done

File: util/crypto/crypto.go
Comment (crenshaw-dev) [inline util/crypto/crypto.go:16]: s/if/is ?

**Review by crenshaw-dev — CHANGES_REQUESTED** [review]

File: util/crypto/crypto.go
Comment (alexmt) [reply util/crypto/crypto.go]: fixed

**Review by crenshaw-dev — APPROVED** [review]
LGTM!

**Pull request merged**
